katas/content/superposition_kata/all_bell_states/Verification.qs

```diff
@@ -0,0 +1,38 @@
+namespace Kata.Verification {
+    open Microsoft.Quantum.Katas;
+
+    operation AllBellStates (qs : Qubit[], index : Int) : Unit is Adj + Ctl {
+        H(qs[0]);
+        
+        if index == 1 {
+            Z(qs[0]);
+        }
+        if index == 2 {
+            X(qs[1]);
+        }
+        if index == 3 {
+            Z(qs[0]);
+            X(qs[1]);
+        }
+        
+        CNOT(qs[0], qs[1]);
+    }
+
+    @EntryPoint()
+    operation CheckSolution() : Bool {
+        for index in 0 .. 3 {
+            if not CheckOperationsEquivalenceOnZeroStateAndIndexWithFeedback(
+                Kata.AllBellStates,
+                AllBellStates,
+                2,
+                index) {
+                Message("Incorrect");
```

Comment:
We can simplify this code using [partial application](https://learn.microsoft.com/en-us/azure/quantum/user-guide/language/expressions/closures#partial-application), and then we won't need the changes to the libraries at all!
```suggestion
            if not CheckOperationsEquivalenceOnZeroStateWithFeedback(
                Kata.AllBellStates(_, index),
                AllBellStates(_, index),
                2) {
```

We'll be using different signatures for the states we want to compare - sometimes integers, other times Booleans or arrays, so it's not really practical to introduce a new library for each signature of the operation we want to test. Instead, we bind the integer parameter of these operations to the index we want to test and leave qubit parameters unbound - the result is an operation with a single parameter that we can pass to a regular library. Partial application is very useful and especially in testing - let me know if you want more details, I'll try to dig up more docs.

(And with this, you can revert the change to the KatasLibrary.qs)

Reply:
good to know! thanks for the tip.